design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,68 @@
+# Overview
+Today, anything that happens on the host side of VMIs (i.e. virt-launcher) is transient.  
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC for the whole cluster.  
```

Comment:
The idea of 1 PVC per VM was brought up at the KubeVirt community meeting, and discarded for its high disk usage.
It was stated that storage backends tend to allocated a minimum of 1GiB per PV, which would be a massive waste of space if there was 1 per VM...

Reply:
@jean-edouard this is a very vague idea, but why don't we use [csi-hostpath](https://github.com/kubevirt/hostpath-provisioner) for doing a proper partitioning? At least this CSI driver is something maintained by KubeVirt community. 

@awels do we have a minimum size for the PVC generated by the hostprovisioner?

Reply:
So the hostpath provisioner generates PVs with capacity of the total size, which should match most PVC request sizes (note the difference between capacity and request sizes). Since the hostpath provisioner does not have anyway of enforcing size it just reports the total size (of which not everything will be available). The only provisioner I am aware of that allocates in PVs in 1Gi chunks is the cinder provisioner. ceph allows smaller sizes as well.

Reply:
@awels that sounds good! In that case, we should be able to just create 1 per VM, simplifying things a lot. On migration, I assume we can move the hostpath to the target node?

Reply:
I am missing a piece here, when you say migration, you don't mean live migration, since that is not possible without RWX storage.

Reply:
Okay so if you want to live migrate a VM, you need RWX storage so it can be mounted on 2 nodes at once. RWO by definition cannot be mounted on 2 nodes at once. hostpath provisioner kind of by definition can only generate RWO volumes since it is just creating directories on a host. So you cannot live migrate.

Reply:
It's important to note that the storage is only for virt-launcher, it's not exposed to the guest, and libvirt doesn't know it exists.
So, theoretically, we can handle it however we want, like in a way that never has it mounted on 2 nodes at once...
The simplest thing would be to make a duplicate of it on migration, if that's a possible operation.

Reply:
> It's important to note that the storage is only for virt-launcher, it's not exposed to the guest, and libvirt doesn't know it exists. So, theoretically, we can handle it however we want, like in a way that never has it mounted on 2 nodes at once... The simplest thing would be to make a duplicate of it on migration, if that's a possible operation.

It will be interesting to understand how libvirt treats the vTPMs during migration. If it is copied over the network like local storage (in KubeVirt an example is the delta with containerdisks) or it expects that the data are available also on the destination node. If there is such a mechanism, then you can achieve live migration by creating a second PVC in the destination node and let libvirt copying the information. From your description, the data seems to be relatively small so I guess it shouldn't have a big impact on the performance.

Reply:
Yeah, the file is just copied over. It probably does it while the guest is down to avoid worrying about (unlikely) deltas, and since the file is so small, but I don't know for sure.
Creating a second PVC sounds just fine, we'd just have to figure out how to maintain the link between VM object and PVC.

Reply:
maybe you could set the ownerReference on the PVC to the VM object? If you do so, the PVC is automatically deleted when the VM object is deleted

Reply:
Great idea! Thanks for all the info/ideas, I will revise the proposal soon to include that.